distribution/registry.go

```diff
@@ -16,6 +16,8 @@ import (
 	"github.com/docker/docker/registry"
 	"github.com/docker/engine-api/types"
 	"golang.org/x/net/context"
+
+	middleware "github.com/docker/docker/api/server/middleware"
```

Comment:
@aaronlehmann  Thanks for reviewing this; your comments were awesome.  I implemented your idea about having push/pop methods inside `dockerversion` and not exposing the type.  `middleware` is not imported anymore into `distribution`, and `distribution` itself only knows about `dockerversion` helper methods.

I've done everything you suggestsed except removing the optional arg hack on `DockerUserAgent()`.  Still contemplating that one.  I'm reluctant only because I'd have to change every call to this function from `DockerUserAgent()` to `DockerUserAgent("")`.